File: blueprint/routes/blueprint.ts
/**
 * Blueprint Path Generator - Claude API Integration
 *
 * Generates 3 distinct implementation paths for user ideas.
 * Uses Claude Sonnet 4 to analyze idea + features and recommend approaches.
 */

import Anthropic from "@anthropic-ai/sdk";

const client = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY || ""
});

export interface BlueprintPath {
  id: string;
  name: string;
  tagline: string;
  why_recommended: string;
  pros: string[];
  cons: string[];
  tech_approach: string;
  time_estimate: string;
  complexity: "Simple" | "Moderate" | "Complex";
}

export interface BlueprintData {
  recommended: string;
  paths: BlueprintPath[];
  bee_note: string;
}

export interface BlueprintResult {
  ok: boolean;
  data?: BlueprintData;
  error?: string;
}

const SYSTEM_PROMPT = `You are Blueprint, Worker Bee's strategic planning mode.

Your job: Read the user's core idea and must-have features, then generate exactly 3 distinct implementation paths.

CRITICAL RULES:
1. Respond ONLY with valid JSON — no markdown, no preamble, no explanation, nothing outside the JSON object
2. Exactly 3 paths, ordered from simplest to most complex
3. Paths must be genuinely different approaches, not just variations in scope
4. One path MUST be recommended (set "recommended" to 0, 1, or 2)
5. Be specific to THIS idea — no generic advice
6. Complexity must be exactly "Simple", "Moderate", or "Complex"
7. Time estimates must be realistic (format: "2-3 hours", "1-2 days", "1 week", etc.)
8. Pros and cons must be specific to the idea, not generic platitudes

JSON SCHEMA (return exactly this structure):
{
  "recommended": "0",
  "paths": [
    {
      "id": "0",
      "name": "Quick & Minimal",
      "tagline": "One sentence describing this approach",
      "why_recommended": "2-3 sentences explaining why this path makes sense FOR THIS SPECIFIC IDEA",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "2-3 sentences describing the tech stack and architecture",
      "time_estimate": "2-3 hours",
      "complexity": "Simple"
    },
    {
      "id": "1",
      "name": "Balanced Build",
      "tagline": "One sentence describing this approach",
      "why_recommended": "2-3 sentences explaining why this path makes sense FOR THIS SPECIFIC IDEA",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "2-3 sentences describing the tech stack and architecture",
      "time_estimate": "1-2 days",
      "complexity": "Moderate"
    },
    {
      "id": "2",
      "name": "Production-Ready",
      "tagline": "One sentence describing this approach",
      "why_recommended": "2-3 sentences explaining why this path makes sense FOR THIS SPECIFIC IDEA",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "2-3 sentences describing the tech stack and architecture",
      "time_estimate": "1-2 weeks",
      "complexity": "Complex"
    }
  ],
  "bee_note": "One honest sentence from Worker Bee about the most important thing to get right with this idea"
}

TONE: You're a warm, experienced colleague who's built things before. Be honest about tradeoffs. Don't oversell or undersell. Help them pick the right path for their context.

REMEMBER: Return ONLY the JSON object. No markdown code blocks. No explanations. Just pure JSON.`;

export async function generateBlueprintPaths(
  idea: string,
  features: string[]
): Promise<BlueprintResult> {
  // Validate API key
  if (!process.env.ANTHROPIC_API_KEY) {
    return {
      ok: false,
      error: "ANTHROPIC_API_KEY not configured. Add it to ~/worker-bee/.env"
    };
  }

  // Validate inputs
  if (!idea || idea.trim().length === 0) {
    return {
      ok: false,
      error: "Core idea cannot be empty"
    };
  }

  try {
    // Format features as bullet list
    const featuresText =
      features && features.length > 0
        ? features.map((f) => `- ${f.trim()}`).join("\n")
        : "- No specific features listed";

    const userPrompt = `Project Idea:
${idea.trim()}

Must-Have Features:
${featuresText}

Generate 3 distinct implementation paths for this idea. Return only the JSON object.`;

    // Call Claude API
    const response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 4096,
      system: SYSTEM_PROMPT,
      messages: [
        {
          role: "user",
          content: userPrompt
        }
      ]
    });

    // Extract response text
    const firstBlock = response.content[0];
    if (firstBlock.type !== "text") {
      return {
        ok: false,
        error: "Claude returned non-text response"
      };
    }
    const responseText = firstBlock.text.trim();

    // Remove markdown code blocks if present (Claude sometimes adds them despite instructions)
    let cleanedResponse = responseText;
    if (responseText.startsWith("```")) {
      const lines = responseText.split("\n");
      const jsonLines: string[] = [];
      let inCodeBlock = false;

      for (const line of lines) {
        if (line.startsWith("```")) {
          inCodeBlock = !inCodeBlock;
          continue;
        }
        if (inCodeBlock || !line.startsWith("```")) {
          jsonLines.push(line);
        }
      }
      cleanedResponse = jsonLines.join("\n").trim();
    }

    // Parse JSON response
    let data: any;
    try {
      data = JSON.parse(cleanedResponse);
    } catch (parseError) {
      return {
        ok: false,
        error: `Claude returned invalid JSON. Parse error: ${parseError instanceof Error ? parseError.message : String(parseError)}\n\nResponse preview: ${cleanedResponse.substring(0, 200)}...`
      };
    }

    // Validate schema
    if (typeof data !== "object" || data === null) {
      return {
        ok: false,
        error: "Claude response is not a valid JSON object"
      };
    }

    if (!("paths" in data)) {
      return {
        ok: false,
        error: "Claude response missing 'paths' field"
      };
    }

    if (!("recommended" in data)) {
      return {
        ok: false,
        error: "Claude response missing 'recommended' field"
      };
    }

    if (!Array.isArray(data.paths)) {
      return {
        ok: false,
        error: "'paths' field must be an array"
      };
    }

    if (data.paths.length !== 3) {
      return {
        ok: false,
        error: `Expected exactly 3 paths, got ${data.paths.length}`
      };
    }

    // Validate recommended index
    if (
      typeof data.recommended !== "string" ||
      !["0", "1", "2"].includes(data.recommended)
    ) {
      return {
        ok: false,
        error: `Invalid 'recommended' value: ${data.recommended} (must be "0", "1", or "2")`
      };
    }

    // Validate each path has required fields
    const requiredFields = [
      "id",
      "name",
      "tagline",
      "why_recommended",
      "pros",
      "cons",
      "tech_approach",
      "time_estimate",
      "complexity"
    ];

    for (let i = 0; i < data.paths.length; i++) {
      const path = data.paths[i];

      for (const field of requiredFields) {
        if (!(field in path)) {
          return {
            ok: false,
            error: `Path ${i} missing required field: '${field}'`
          };
        }
      }

      // Validate complexity is one of the allowed values
      if (!["Simple", "Moderate", "Complex"].includes(path.complexity)) {
        return {
          ok: false,
          error: `Path ${i} has invalid complexity: '${path.complexity}' (must be 'Simple', 'Moderate', or 'Complex')`
        };
      }

      // Validate pros and cons are arrays
      if (!Array.isArray(path.pros)) {
        return {
          ok: false,
          error: `Path ${i} 'pros' must be an array`
        };
      }

      if (!Array.isArray(path.cons)) {
        return {
          ok: false,
          error: `Path ${i} 'cons' must be an array`
        };
      }

      // Validate id matches index (as string)
      if (path.id !== String(i)) {
        return {
          ok: false,
          error: `Path ${i} has mismatched id: ${path.id} (should be "${i}")`
        };
      }
    }

    // Validate bee_note exists
    if (!("bee_note" in data) || typeof data.bee_note !== "string") {
      return {
        ok: false,
        error: "Missing or invalid 'bee_note' field"
      };
    }

    // All validation passed - return successful result
    return {
      ok: true,
      data: data as BlueprintData
    };

  } catch (error) {
    // Handle API errors
    if (error instanceof Anthropic.APIError) {
      return {
        ok: false,
        error: `Claude API error: ${error.message}`
      };
    }

    // Handle unexpected errors
    return {
      ok: false,
      error: `Unexpected error: ${error instanceof Error ? error.message : String(error)}`
    };
  }
}

File: agent/tools/blueprint.py
"""
Blueprint Tool - Claude API integration for path recommendations

Generates 3 structured implementation paths for user ideas.
Returns JSON with recommended path, pros/cons, tech approach, complexity.
"""

import os
import json
from typing import Dict, List, Any
import anthropic


async def generate_blueprint_paths(idea: str, features: List[str]) -> Dict[str, Any]:
    """
    Generate 3 implementation path recommendations for a given idea.

    Args:
        idea: Core idea description from user
        features: List of must-have feature bullet points

    Returns:
        {
            "ok": bool,
            "data": {
                "recommended": int,  # 0, 1, or 2 - which path is recommended
                "paths": [
                    {
                        "id": int,
                        "name": str,
                        "tagline": str,
                        "why_recommended": str,
                        "pros": List[str],
                        "cons": List[str],
                        "tech_approach": str,
                        "time_estimate": str,
                        "complexity": str  # "Simple", "Moderate", or "Complex"
                    }
                ],
                "bee_note": str  # Personal note from Worker Bee
            },
            "error": str  # Only present if ok=False
        }
    """

    api_key = os.getenv("ANTHROPIC_API_KEY", "")
    if not api_key:
        return {
            "ok": False,
            "error": "ANTHROPIC_API_KEY not found in environment"
        }

    try:
        client = anthropic.Anthropic(api_key=api_key)

        # Format features as bullet list
        features_list = "\n".join([f"- {f}" for f in features]) if features else "No specific features listed"

        # System prompt forces JSON-only response with strict schema
        system_prompt = """You are Worker Bee's path recommendation engine.

Given a user's core idea and must-have features, generate exactly 3 implementation paths.

CRITICAL RULES:
1. Return ONLY valid JSON - no markdown, no explanations, just the JSON object
2. Exactly 3 paths, ordered from simplest to most complex
3. One path MUST be recommended (set recommended: 0, 1, or 2)
4. Complexity must be exactly "Simple", "Moderate", or "Complex"
5. Time estimates realistic (hours/days/weeks format)
6. Pros and cons must be specific, not generic
7. Tech approach describes the stack and architecture briefly

JSON SCHEMA (return this exact structure):
{
  "recommended": 0,
  "paths": [
    {
      "id": 0,
      "name": "Quick & Minimal",
      "tagline": "Get it working fast with essential features only",
      "why_recommended": "Why this path makes sense for this specific idea",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "Brief description of stack and architecture",
      "time_estimate": "1-2 days",
      "complexity": "Simple"
    },
    {
      "id": 1,
      "name": "Balanced Build",
      "tagline": "Solid foundation with room to grow",
      "why_recommended": "Why this path makes sense for this specific idea",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "Brief description of stack and architecture",
      "time_estimate": "3-5 days",
      "complexity": "Moderate"
    },
    {
      "id": 2,
      "name": "Production-Ready",
      "tagline": "Enterprise-grade with all the features",
      "why_recommended": "Why this path makes sense for this specific idea",
      "pros": ["Specific advantage 1", "Specific advantage 2", "Specific advantage 3"],
      "cons": ["Specific limitation 1", "Specific limitation 2"],
      "tech_approach": "Brief description of stack and architecture",
      "time_estimate": "1-2 weeks",
      "complexity": "Complex"
    }
  ],
  "bee_note": "Personal note from Worker Bee about this idea (1-2 sentences, warm colleague tone)"
}

REMEMBER: Return ONLY the JSON object. No markdown code blocks, no explanations."""

        user_prompt = f"""Core Idea:
{idea}

Must-Have Features:
{features_list}

Generate 3 implementation paths for this idea. Return only the JSON object."""

        response = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=4096,
            system=system_prompt,
            messages=[{"role": "user", "content": user_prompt}]
        )

        # Extract JSON from response
        response_text = response.content[0].text.strip()

        # Remove markdown code blocks if present (Claude sometimes adds them despite instructions)
        if response_text.startswith("```"):
            # Find the actual JSON content between code blocks
            lines = response_text.split("\n")
            json_lines = []
            in_code_block = False
            for line in lines:
                if line.startswith("```"):
                    in_code_block = not in_code_block
                    continue
                if in_code_block or not line.startswith("```"):
                    json_lines.append(line)
            response_text = "\n".join(json_lines).strip()

        # Parse JSON response
        try:
            data = json.loads(response_text)
        except json.JSONDecodeError as e:
            return {
                "ok": False,
                "error": f"Claude returned invalid JSON: {str(e)}\nResponse: {response_text[:200]}"
            }

        # Validate schema
        if "paths" not in data or "recommended" not in data:
            return {
                "ok": False,
                "error": "Claude response missing required fields (paths, recommended)"
            }

        if len(data["paths"]) != 3:
            return {
                "ok": False,
                "error": f"Expected 3 paths, got {len(data['paths'])}"
            }

        # Validate each path has required fields
        required_fields = ["id", "name", "tagline", "why_recommended", "pros", "cons", "tech_approach", "time_estimate", "complexity"]
        for i, path in enumerate(data["paths"]):
            for field in required_fields:
                if field not in path:
                    return {
                        "ok": False,
                        "error": f"Path {i} missing required field: {field}"
                    }

            # Validate complexity is one of the allowed values
            if path["complexity"] not in ["Simple", "Moderate", "Complex"]:
                return {
                    "ok": False,
                    "error": f"Path {i} has invalid complexity: {path['complexity']} (must be Simple, Moderate, or Complex)"
                }

        # Validate recommended index
        if not isinstance(data["recommended"], int) or data["recommended"] not in [0, 1, 2]:
            return {
                "ok": False,
                "error": f"Invalid recommended index: {data['recommended']} (must be 0, 1, or 2)"
            }

        return {
            "ok": True,
            "data": data
        }

    except anthropic.APIError as e:
        return {
            "ok": False,
            "error": f"Anthropic API error: {str(e)}"
        }
    except Exception as e:
        return {
            "ok": False,
            "error": f"Unexpected error: {str(e)}"
        }
